TLS connection errors sometimes occur

I am using vercel and sending the request to algolia via vercel.

When I use search-instights.js I get the following error and cannot send a click event.
It does not occur every time, but once every few times.
Also, this does not occur when running at local.

```
2023-12-04T15:48:24.883Z	625e31bf-bc53-49fb-8835-ff936d6e5078	ERROR	Error: Client network socket disconnected before secure TLS connection was established
    at connResetException (node:internal/errors:720:14)
    at TLSSocket.onConnectEnd (node:_tls_wrap:1655:19)
    at TLSSocket.emit (node:events:529:35)
    at endReadableNT (node:internal/streams/readable:1368:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  code: 'ECONNRESET',
  path: null,
  host: 'insights.algolia.io',
  port: 443,
  localAddress: undefined
}
2023-12-04T15:48:24.884Z	625e31bf-bc53-49fb-8835-ff936d6e5078	ERROR	Unhandled Promise Rejection 
{
   "errorType":"Runtime.UnhandledPromiseRejection",
   "errorMessage":"Error: Client network socket disconnected before secure TLS connection was established",
   "reason":{
      "errorType":"Error",
      "errorMessage":"Client network socket disconnected before secure TLS connection was established",
      "code":"ECONNRESET",
      "path":null,
      "host":"insights.algolia.io",
      "port":443,
      "stack":[
         "Error: Client network socket disconnected before secure TLS connection was established",
         "    at connResetException (node:internal/errors:720:14)",
         "    at TLSSocket.onConnectEnd (node:_tls_wrap:1655:19)",
         "    at TLSSocket.emit (node:events:529:35)",
         "    at endReadableNT (node:internal/streams/readable:1368:12)",
         "    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)"
      ]
   },
   "promise":{
      
   },
   "stack":[
      "Runtime.UnhandledPromiseRejection: Error: Client network socket disconnected before secure TLS connection was established",
      "    at process.<anonymous> (file:///var/runtime/index.mjs:1276:17)",
      "    at process.emit (node:events:529:35)",
      "    at emit (node:internal/process/promises:149:20)",
      "    at processPromiseRejections (node:internal/process/promises:283:27)",
      "    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"
   ]
}
[ERROR] [1701704904888] LAMBDA_RUNTIME Failed to post handler success response. Http response code: 400.
Error: Runtime exited with error: exit status 128
Runtime.ExitError
```

Here is the code.
https://github.com/mongolyy/next-with-algolia-app

Since Vercel is running on Lambda and there was a LAMBDA_RUNTIME error message in this log as well, I believe the following issue may be related.
https://github.com/aws/aws-sdk-js/issues/3591
I think I might need to await for the execution of some asynchronous function, but I couldn't figure out how to do it.

Any suggestion?

--- a/lib/__tests__/_getFunctionalInterface.test.ts
+++ b/lib/__tests__/_getFunctionalInterface.test.ts
@@ -2,7 +2,11 @@
 
 import { getFunctionalInterface } from "../_getFunctionalInterface";
 import AlgoliaAnalytics from "../insights";
-import type { InsightsClient } from "../types";
+import type { InsightsClient, InsightsMethodMap } from "../types";
+
+jest.mock("../../package.json", () => ({
+  version: "1.0.1"
+}));
 
 describe("_getFunctionalInterface", () => {
   let aa: InsightsClient;
@@ -22,5 +26,63 @@ describe("_getFunctionalInterface", () => {
     expect(console.warn).toHaveBeenCalledWith(
       "The method `unknown-function` doesn't exist."
     );
+  });
+
+  it.each<keyof InsightsMethodMap>([
+    "addedToCartObjectIDs",
+    "addedToCartObjectIDsAfterSearch",
+    "clickedFilters",
+    "clickedObjectIDs",
+    "clickedObjectIDsAfterSearch",
+    "convertedFilters",
+    "convertedObjectIDs",
+    "convertedObjectIDsAfterSearch",
+    "purchasedObjectIDs",
+    "purchasedObjectIDsAfterSearch",
+    "viewedFilters",
+    "viewedObjectIDs"
+  ])("returns a promise to be waited on for %s call", async (method) => {
+    aa("init", {
+      appId: "appId",
+      apiKey: "apiKey"
+    });
+    const result = await aa(method, {
+      eventName: "click",
+      objectIDs: ["1", "2"],
+      index: "index"
+    });
+    expect(result).toBe(true);
+  });
+
+  it("returns a promise to be waited on for sendEvents call", async () => {
+    aa("init", {
+      appId: "appId",
+      apiKey: "apiKey"
+    });
+    const result = await aa("sendEvents", [
+      {
+        eventName: "click",
+        eventType: "click",
+        objectIDs: ["1", "2"],
+        index: "index"
+      }
+    ]);
+    expect(result).toBe(true);
+  });
+
+  it("returns values on sync calls", () => {
+    aa("init", {
+      appId: "appId",
+      apiKey: "apiKey",
+      userToken: "userToken",
+      authenticatedUserToken: "authenticatedUserToken"
+    });
+    expect(aa("getAuthenticatedUserToken")).toBe("authenticatedUserToken");
+    expect(aa("getUserToken")).toBe("userToken");
+    expect(aa("getVersion")).toBe("1.0.1");
+    expect(aa("setAuthenticatedUserToken", "newAuthenticatedUserToken")).toBe(
+      "newAuthenticatedUserToken"
+    );
+    expect(aa("setUserToken", "newUserToken")).toBe("newUserToken");
   });
 });

--- a/lib/_getFunctionalInterface.ts
+++ b/lib/_getFunctionalInterface.ts
@@ -6,12 +6,12 @@ export function getFunctionalInterface(
   instance: AlgoliaAnalytics
 ): InsightsClient {
   return (functionName, ...functionArguments) => {
-    if (functionName && isFunction((instance as any)[functionName])) {
+    if (functionName && isFunction(instance[functionName])) {
       // @ts-expect-error
-      instance[functionName](...functionArguments);
-    } else {
-      // eslint-disable-next-line no-console
-      console.warn(`The method \`${functionName}\` doesn't exist.`);
+      return instance[functionName](...functionArguments);
     }
+    // eslint-disable-next-line no-console
+    console.warn(`The method \`${functionName}\` doesn't exist.`);
+    return undefined;
   };
 }

--- a/lib/_getVersion.ts
+++ b/lib/_getVersion.ts
@@ -3,9 +3,10 @@ import { isFunction } from "./utils";
 
 export function getVersion(
   this: AlgoliaAnalytics,
-  callback: (version: string) => void
-): void {
+  callback?: (version: string) => void
+): string {
   if (isFunction(callback)) {
     callback(this.version);
   }
+  return this.version;
 }

--- a/lib/_sendEvent.ts
+++ b/lib/_sendEvent.ts
@@ -9,9 +9,9 @@ export function makeSendEvents(requestFn: RequestFnType) {
     this: AlgoliaAnalytics,
     eventData: InsightsEvent[],
     additionalParams?: InsightsAdditionalEventParams
-  ): Promise<boolean> | undefined {
+  ): Promise<boolean> {
     if (this._userHasOptedOut) {
-      return undefined;
+      return Promise.resolve(false);
     }
     const hasCredentials =
       (!isUndefined(this._apiKey) && !isUndefined(this._appId)) ||

--- a/lib/types.ts
+++ b/lib/types.ts
@@ -26,43 +26,53 @@ import type {
 import type { init } from "./init";
 import type { viewedObjectIDs, viewedFilters } from "./view";
 
+type ParamReturnTypeTuple<T extends (...args: any) => any> = [
+  Parameters<T>,
+  ReturnType<T>
+];
 export type InsightsMethodMap = {
-  init: Parameters<typeof init>;
-  getVersion: Parameters<typeof getVersion>;
-  addAlgoliaAgent: Parameters<typeof addAlgoliaAgent>;
-  setUserToken: Parameters<typeof setUserToken>;
-  getUserToken: Parameters<typeof getUserToken>;
-  onUserTokenChange: Parameters<typeof onUserTokenChange>;
-  setAuthenticatedUserToken: Parameters<typeof setAuthenticatedUserToken>;
-  getAuthenticatedUserToken: Parameters<typeof getAuthenticatedUserToken>;
-  onAuthenticatedUserTokenChange: Parameters<
+  init: ParamReturnTypeTuple<typeof init>;
+  getVersion: ParamReturnTypeTuple<typeof getVersion>;
+  addAlgoliaAgent: ParamReturnTypeTuple<typeof addAlgoliaAgent>;
+  setUserToken: ParamReturnTypeTuple<typeof setUserToken>;
+  getUserToken: ParamReturnTypeTuple<typeof getUserToken>;
+  onUserTokenChange: ParamReturnTypeTuple<typeof onUserTokenChange>;
+  setAuthenticatedUserToken: ParamReturnTypeTuple<
+    typeof setAuthenticatedUserToken
+  >;
+  getAuthenticatedUserToken: ParamReturnTypeTuple<
+    typeof getAuthenticatedUserToken
+  >;
+  onAuthenticatedUserTokenChange: ParamReturnTypeTuple<
     typeof onAuthenticatedUserTokenChange
   >;
-  clickedObjectIDsAfterSearch: Parameters<typeof clickedObjectIDsAfterSearch>;
-  clickedObjectIDs: Parameters<typeof clickedObjectIDs>;
-  clickedFilters: Parameters<typeof clickedFilters>;
-  convertedObjectIDsAfterSearch: Parameters<
+  clickedObjectIDsAfterSearch: ParamReturnTypeTuple<
+    typeof clickedObjectIDsAfterSearch
+  >;
+  clickedObjectIDs: ParamReturnTypeTuple<typeof clickedObjectIDs>;
+  clickedFilters: ParamReturnTypeTuple<typeof clickedFilters>;
+  convertedObjectIDsAfterSearch: ParamReturnTypeTuple<
     typeof convertedObjectIDsAfterSearch
   >;
-  convertedObjectIDs: Parameters<typeof convertedObjectIDs>;
-  convertedFilters: Parameters<typeof convertedFilters>;
-  viewedObjectIDs: Parameters<typeof viewedObjectIDs>;
-  viewedFilters: Parameters<typeof viewedFilters>;
-  purchasedObjectIDs: Parameters<typeof purchasedObjectIDs>;
-  purchasedObjectIDsAfterSearch: Parameters<
+  convertedObjectIDs: ParamReturnTypeTuple<typeof convertedObjectIDs>;
+  convertedFilters: ParamReturnTypeTuple<typeof convertedFilters>;
+  viewedObjectIDs: ParamReturnTypeTuple<typeof viewedObjectIDs>;
+  viewedFilters: ParamReturnTypeTuple<typeof viewedFilters>;
+  purchasedObjectIDs: ParamReturnTypeTuple<typeof purchasedObjectIDs>;
+  purchasedObjectIDsAfterSearch: ParamReturnTypeTuple<
     typeof purchasedObjectIDsAfterSearch
   >;
-  addedToCartObjectIDs: Parameters<typeof addedToCartObjectIDs>;
-  addedToCartObjectIDsAfterSearch: Parameters<
+  addedToCartObjectIDs: ParamReturnTypeTuple<typeof addedToCartObjectIDs>;
+  addedToCartObjectIDsAfterSearch: ParamReturnTypeTuple<
     typeof addedToCartObjectIDsAfterSearch
   >;
-  sendEvents: Parameters<ReturnType<typeof makeSendEvents>>;
+  sendEvents: ParamReturnTypeTuple<ReturnType<typeof makeSendEvents>>;
 };
 
 type MethodType<MethodName extends keyof InsightsMethodMap> = (
   method: MethodName,
-  ...args: InsightsMethodMap[MethodName]
-) => void;
+  ...args: InsightsMethodMap[MethodName][0]
+) => InsightsMethodMap[MethodName][1];
 
 export type Init = MethodType<"init">;
 
@@ -98,8 +108,8 @@ export type SendEvents = MethodType<"sendEvents">;
 
 export type InsightsClient = (<MethodName extends keyof InsightsMethodMap>(
   method: MethodName,
-  ...args: InsightsMethodMap[MethodName]
-) => void) & { version?: string };
+  ...args: InsightsMethodMap[MethodName][0]
+) => InsightsMethodMap[MethodName][1]) & { version?: string };
 
 export type InsightsEventType = "click" | "conversion" | "view";
 export type InsightsEventConversionSubType = "addToCart" | "purchase";

